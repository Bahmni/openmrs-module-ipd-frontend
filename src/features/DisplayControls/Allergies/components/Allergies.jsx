import {
  DataTable,
  DataTableSkeleton,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
} from "carbon-components-react";
import React, { useEffect, useState } from "react";
import { useFetchAllergiesIntolerance } from "../hooks/useFetchAllergiesIntolerance";
import PropTypes from "prop-types";
import "../styles/Allergies.scss";
import { FormattedMessage } from "react-intl";

const Allergies = (props) => {
  const { patientId } = props;

  const { allergiesData, isLoading } = useFetchAllergiesIntolerance(patientId);
  const [rows, setRows] = useState([]);
  const NoAllergenMessage = (
    <FormattedMessage
      id={"NO_ALLERGENS_MESSAGE"}
      defaultMessage={"No Allergen is captured for this patient yet."}
    />
  );

  useEffect(() => {
    if (allergiesData && allergiesData.entry) {
      const allergies = [];
      allergiesData.entry?.map((allergy) => {
        allergies.push({
          allergen: allergy.resource.code.coding[0].display,
          id: allergy.resource.id,
          severity: getSeverity(allergy.resource.criticality),
          reaction: getAllergyReactions(allergy.resource.reaction),
          comments: getComments(allergy.resource.note),
          sortWeight: getSortingWait(getSeverity(allergy.resource.criticality)),
        });
      });
      setRows(sortedRow(allergies));
    }
  }, [allergiesData]);

  const getSortingWait = (severity) => {
    if (severity === "Severe") return -1;
    if (severity === "Moderate") return 0;
    return 1;
  };

  const getSeverity = (criticality) => {
    if (criticality == "unable-to-assess") return "Moderate";
    else if (criticality == "high") return "Severe";
    else return "Mild";
  };

  const getComments = (notes) =>
    notes && notes.length > 0 ? notes[0].text : "";

  const getAllergyReactions = (reactions) => {
    let allergyReactions = "";
    if (reactions && reactions.length > 0) {
      reactions[0].manifestation.map((reaction) => {
        allergyReactions =
          allergyReactions != ""
            ? `${allergyReactions}, ${reaction.coding[0].display}`
            : `${reaction.coding[0].display}`;
      });
    }
    return allergyReactions;
  };

  const headers = [
    {
      key: "allergen",
      header: "Allergen",
    },
    {
      key: "severity",
      header: "Severity",
    },
    {
      key: "reaction",
      header: "Reaction",
    },
    {
      key: "comments",
      header: "Comments",
    },
  ];

  const sortedRow = (rows) => {
    const sortedRows = rows
      .sort((a, b) => {
        if (a === b) return 0;
        return a.allergen > b.allergen ? 1 : -1;
      })
      .sort((a, b) => a.sortWeight - b.sortWeight);
    return sortedRows;
  };

  if (isLoading)
    return (
      <DataTableSkeleton
        data-testid="datatable-skeleton"
        headers={headers}
        aria-label="sample table"
        zebra="true"
      />
    );

  return allergiesData.entry === undefined ? (
    <div className="no-allergen-message"> {NoAllergenMessage} </div>
  ) : (
    <DataTable
      rows={rows}
      headers={headers}
      useZebraStyles={true}
      data-testid="datatable"
    >
      {({ rows, headers, getTableProps, getHeaderProps, getRowProps }) => (
        <Table {...getTableProps()} useZebraStyles>
          <TableHead>
            <TableRow>
              {headers.map((header, index) => (
                <TableHeader
                  key={index + header.key}
                  {...getHeaderProps({ header })}
                  isSortable={header.key === "severity"}
                >
                  {header.header}
                </TableHeader>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {rows.map((row, index) => (
              <TableRow
                key={index + row.id}
                {...getRowProps({ row })}
                data-testid="table-body-row"
              >
                {row.cells.map((cell) => (
                  <TableCell
                    key={cell.id}
                    className={
                      cell.id.includes("severity") && cell.value == "Severe"
                        ? "high-severity-color"
                        : ""
                    }
                  >
                    {cell.value}
                  </TableCell>
                ))}
              </TableRow>
            ))}
          </TableBody>
        </Table>
      )}
    </DataTable>
  );
};

Allergies.propTypes = {
  patientId: PropTypes.string.isRequired,
};

export default Allergies;
