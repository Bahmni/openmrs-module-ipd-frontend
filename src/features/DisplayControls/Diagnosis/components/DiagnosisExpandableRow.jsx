import React from "react";
import PropTypes from "prop-types";
import "../styles/DiagnosisExpandableRow.scss";

const DiagnosisExpandableRow = (props) => {
  const { data } = props;
  console.log("data inside DiagnosisExpandableRow", data);
  return (
    <>
      <h6 className="notes-heading">Notes</h6>
      <div className="notes-content">
        {data.diagnosisNotes ? data.diagnosisNotes : "No notes available"}
      </div>
      {data.diagnosisNotes !== "" && (
        <div className="notes-provider-details">
          {data?.provider} | {data.diagnosisTime}
        </div>
      )}
    </>
  );
};

DiagnosisExpandableRow.propTypes = {
  data: PropTypes.object,
};
export default DiagnosisExpandableRow;
