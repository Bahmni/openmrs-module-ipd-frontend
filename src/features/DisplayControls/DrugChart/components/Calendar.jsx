import React from "react";
import PropTypes from "prop-types";
import CalendarRow from "./CalendarRow.jsx";
import "../styles/Calendar.scss";
export default function Calendar(props) {
  const { calendarData, currentShiftArray, selectedDate } = props;
  return (
    <table className="drug-chart-calendar">
      <tbody>
        {calendarData.map((rowData, index) => {
          return (
            <tr key={index}>
              <CalendarRow
                rowData={rowData}
                currentShiftArray={currentShiftArray}
                selectedDate={selectedDate}
              />
            </tr>
          );
        })}
      </tbody>
    </table>
  );
}

Calendar.propTypes = {
  calendarData: PropTypes.array.isRequired,
  currentShiftArray: PropTypes.array,
  selectedDate: PropTypes.instanceOf(Date),
};
