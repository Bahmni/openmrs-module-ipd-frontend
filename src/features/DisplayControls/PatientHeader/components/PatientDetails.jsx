import React from 'react'
import { Tile, Row, Column} from "carbon-components-react";
import PropTypes from 'prop-types';
import { useState } from 'react';
import { useEffect } from 'react';

const PatientDetails = ({patientDetails, patientDetailsHeaders, contacts, relationships}) => {
    const [locationComponent, setLocationComponents]= useState([]);
    const locationMapping = ()=> {
        let reveresdAddress = [];
        for (const key in patientDetails.locations) {
            if (patientDetails.locations.hasOwnProperty(key)) {
              let value =  <span className="details-value">
              {key} : {patientDetails.locations[key] !=null ? patientDetails.locations[key] : " "}
           </span>
         setLocationComponents( (locationComponent) => [...locationComponent, value])
            }
    }}
    useEffect(()=>{
        locationMapping();
    },[])
  return (
    <Tile className="patient-details">
        <Row>
          <Column>
           <Tile className="details-tile">
             <span className="details-headers">
              {patientDetailsHeaders.address}
             </span>
           {
            locationComponent.map((location)=>{
             return location
            })
            }
           </Tile>
          </Column>
          <Column>
            <Tile className="details-tile">
                <span className="details-headers">
                    {patientDetailsHeaders.contactDetails}
                </span>
                {contacts.map((contact)=> { 
                    return (
                        <span key={contact.id} className='details-value'>
                            {contact.name} : {contact.value}
                        </span>
                    )
                })}
           </Tile>
          </Column>
          <Column>
            <Tile className="details-tile">
                <span className="details-headers">
                    {patientDetailsHeaders.relationships}
                </span>
                {relationships.map((relationship)=> { 
                    return (
                        <span key={relationship.id} className='details-value'>
                            {relationship.relationshipType} : {relationship.name}
                        </span>
                    )
                })}
           </Tile>
          </Column>
        </Row>
       </Tile>
  )
};

export default PatientDetails

PatientDetails.propTypes = {
    patientDetails: PropTypes.object.isRequired,
    patientDetailsHeaders: PropTypes.object.isRequired,
    contacts: PropTypes.array.isRequired,
    relationships: PropTypes.array.isRequired
    };