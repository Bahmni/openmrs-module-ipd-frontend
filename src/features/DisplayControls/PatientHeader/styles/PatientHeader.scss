.patient-header {
  margin: 10px 0 10px 0;
  padding: 0;
  .patient-name {
    color: #393939;
    font-size: 20px;
    margin: 10px 0 10px 0;
  }
  .patient-details {
    align-items: center;
  }

  .patient-image {
    width: 56px;
    height: 56px;
    background-color: #007d79;
  }

  .patient-header-loading {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100%;
  }
  .patient-info {
    padding: 0 10px;
    min-width: fit-content;
    color: #525252;
    font-size: 14px;
  }
}

.is-loading {
  height: 64px;
}

.other-info {
  display: flex;
  flex-direction: row;
  justify-content: start;
  flex-wrap: wrap;
  align-items: center;
  padding: 0 0 0 10px;
}

.bx--row.patient-image-and-details {
  align-items: center;
}
