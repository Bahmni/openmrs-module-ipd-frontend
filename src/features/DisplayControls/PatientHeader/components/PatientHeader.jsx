import React, { useEffect, useState }  from "react";
import PropTypes from "prop-types";
import { getGender, mapContact,getConfigsForPatientContactDetails , fetchPatientProfile, mapRelationships, fetchAddressMapping } from "../utils/PatientHeaderUtils";
import { Tile, Grid, Row, Column, SkeletonText,Button, Link } from "carbon-components-react";
import { formatDateAsString } from "../../../../utils/DateFormatter";
import { FormattedMessage } from "react-intl";
import { DDMMYYY_DATE_FORMAT } from "../../../../constants";
import "../styles/PatientHeader.scss";
import { ChevronDown20,ChevronUp20 } from "@carbon/icons-react";
import PatientDetails from "./PatientDetails";

export const PatientHeader = (props) => {
  const { patientId } = props;
  const [showPatientDetails,togglePatientDetails] = useState(false);
  const [patientDetails, updatePatientDetails] = useState({});
  const [isLoading, updateIsLoading] = useState(true);
  const[contacts , setMappedContacts] = useState([]);
  const[relationships , setMappedRelationships] = useState([]);

  const years = <FormattedMessage id="YEARS" defaultMessage="Years" />;
  const showDetails = <FormattedMessage id="SHOW_PATIENT_DETAILS" defaultMessage="Show Details" />;
  const hideDetails = <FormattedMessage id="HIDE_PATIENT_DETAILS" defaultMessage="Hide Details" />;
  const patientDetailsHeaders = {
    address: (
      <FormattedMessage id={"Address"} defaultMessage={"Address"} />
    ),
    contactDetails: (
      <FormattedMessage id={"Contact Details"} defaultMessage={"Contact Details"} />
    ),
    relationships: (
      <FormattedMessage id={"Relationships"} defaultMessage={"Relationships"} />
    )
  }

  const getContactDetailsConfigs = async () => {
    const patientContactConfigs = await getConfigsForPatientContactDetails();
    return patientContactConfigs;
  };

  const extractPatientInfo = (patientInfo, locationMap) => {
    let locations = [];
    updatePatientDetails({
      fullName: patientInfo?.person?.preferredName.display,
      givenName: patientInfo?.person?.preferredName.givenName,
      familyName: patientInfo?.person?.preferredName.familyName,
      middleName: patientInfo?.person?.preferredName?.middleName,
      age: patientInfo?.person?.age,
      birthDate: formatDateAsString(
        new Date(patientInfo?.person?.birthdate),
        DDMMYYY_DATE_FORMAT
      ),
      attributes: patientInfo?.person?.attributes,
      gender: getGender(patientInfo?.person?.gender),
      identifier: patientInfo?.identifiers[0]?.identifier,
    });
    locationMap.map((location) => {
         locations = {...locations,[location.name]: patientInfo.person.preferredAddress && patientInfo.person.preferredAddress[location.addressField]}
      });    
    updatePatientDetails((patientDetails) => ({...patientDetails, locations})
    )  
    updateIsLoading(false);
    return(patientInfo?.person?.attributes);
  };

  const extractPatientRelationships = (patientProfile) => {
  return(patientProfile?.relationships);
}
 
  const toggleDetailsView = () => togglePatientDetails(!showPatientDetails);
  useEffect(() => {
    const getPatientInfo = async () => {
      const patientProfile = await fetchPatientProfile(patientId);
      const patientInfo = patientProfile?.patient;
      const locationMap = await fetchAddressMapping();
      const patientAttributes = extractPatientInfo(patientInfo, locationMap);
      const contactConfigs = await getContactDetailsConfigs();
      const patientRelatives = extractPatientRelationships(patientProfile);
      setMappedContacts(mapContact(patientAttributes , contactConfigs.contactDetails));
      setMappedRelationships(mapRelationships(patientRelatives));
    };
    getPatientInfo();
  }, []);

  return (
    <>
      <Tile className="patient-header">
        {isLoading ? (
          <SkeletonText className="is-loading" data-testid="header-loading" />
        ) : (
          <>
            <Grid>
              <Row className="patient-image-and-details">
                <div className={"patient-image"} />
                <Column>
                  <Row>
                    <Column>
                      <h1 className="patient-name">
                        {patientDetails?.fullName}
                      </h1>
                    </Column>
                  </Row>
                  <Row>
                    <div className="other-info">
                      <div className="patient-basic-info">
                        <h3 className="patient-info">{patientDetails?.gender}</h3>
                        <h3 className="patient-info">
                          {patientDetails?.age} {years}
                        </h3>
                        <h3 className="patient-info">
                          {patientDetails?.birthDate}
                        </h3>
                        <h3 className="patient-info">
                          {patientDetails?.identifier}
                        </h3>
                      </div>
                      {showPatientDetails ? (<Link kind="tertiary" className="show-more" size="sm" onClick={toggleDetailsView}>
                          {hideDetails} <ChevronUp20/> </Link>
                      ) : (
                        <Link kind="tertiary" className="show-more" size="sm" onClick={toggleDetailsView}>
                          {showDetails} <ChevronDown20 /> 
                        </Link>)}
                    </div>
                  </Row>
                </Column>
              </Row>
            </Grid>
            {showPatientDetails && 
              <PatientDetails 
              patientDetails={patientDetails}
              patientDetailsHeaders={patientDetailsHeaders}
              contacts={contacts}
              relationships={relationships}/>}
          </>
        )}
      </Tile>
    </>
  );
};

PatientHeader.propTypes = {
  patientId: PropTypes.string.isRequired,
};
