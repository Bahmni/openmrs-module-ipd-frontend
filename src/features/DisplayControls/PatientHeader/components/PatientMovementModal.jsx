import React, { useState, useContext, useEffect } from "react";
import { IPDContext } from "../../../../context/IPDContext";
import {
  fetchVisitSummary,
  updatePatientMovement,
  dischargePatient,
  undoDischargePatient,
} from "../utils/PatientMovementModalUtils";
import {
  searchConceptsByFSN,
  fetchVisitEncounterOrderTypes,
  getADTDashboardUrl,
} from "../../../../utils/CommonUtils";
import {
  FSN_DISPOSITION_VALUE,
  BY_FSN_VALUE,
  CUSTOM_OUTPUT_VALUE,
  FSN_ADT_NOTES_VALUE,
  BAHMNI_VALUE,
} from "../../../../constants";
import {
  Modal,
  TextArea,
  Dropdown,
  DropdownSkeleton,
} from "carbon-components-react";
import "../styles/PatientHeader.scss";
import PropTypes from 'prop-types';
import { drop } from "lodash";

const PatientMovementModal = (props) => {
  const { updatePatientMovementModal } = props;
  const { patient, visit, location, provider } = useContext(IPDContext);
  const [isLoading, setIsLoading] = useState(true);
  const [dropDown, setDropDown] = useState([]);
  const [visitSummary, setVisitSummary] = useState({});
  const [selectedDropDownItem, setSelectedDropDownItem] = useState({});
  const [currentItem, setCurrentItem] = useState({});
  const [adtNotes, setAdtNotes] = useState();
  const [adtNotesObj, setAdtNotesObj] = useState({});
  const [showAdtNotes, setShowAdtNotes] = useState(false);
  const [saveEnable, setSaveEnable] = useState(false);
  const [vistEncounterTypes, setVisitEncounterTypes] = useState({});
  const fetchDropDownValues = async () => {
    setIsLoading(true);
    const getDropDown = await searchConceptsByFSN(
      BY_FSN_VALUE,
      FSN_DISPOSITION_VALUE,
      CUSTOM_OUTPUT_VALUE
    );
    if (getDropDown.status === 200) {
      const items  = getDropDown.data.results[0].answers;
      var getVisitSummary = await fetchVisitSummary(visit);
      if (getVisitSummary.status === 200 && items.length > 0) {
        const visitSummaryData = getVisitSummary.data;
        setVisitSummary(visitSummaryData);
        var dropDownList = items.map((item) => {
          if (item.name) {
            if (
              visitSummaryData.admissionDetails === null &&
              visitSummaryData.dischargeDetails === null &&
              item.name.name === "Admit Patient"
            ) {
              return { label: item.name.display, value: item.name.display };
            } else if (
              visitSummaryData.admissionDetails !== null &&
              visitSummaryData.dischargeDetails !== null &&
              item.name.name === "Undo Discharge"
            ) {
              return { label: item.name.display, value: item.name.display };
            } else if (
              visitSummaryData.admissionDetails !== null &&
              visitSummaryData.dischargeDetails === null &&
              (item.name.name === "Discharge Patient" ||
                item.name.name === "Transfer Patient")
            ) {
              return { label: item.name.display, value: item.name.display };
            }
          }
        });
        dropDownList = dropDownList.filter(function( element ) {
          return element !== undefined;
       });
        setDropDown(dropDownList);
      }
      var getTypes = await fetchVisitEncounterOrderTypes();
      setVisitEncounterTypes(getTypes.data);
      if (getTypes.data) {
        setIsLoading(false);
      }
    }
  };

  useEffect(() => {
    fetchDropDownValues();
  }, []);

  const fetchAdtNotes = async () => {
    const getAdtNotes = await searchConceptsByFSN(
      BY_FSN_VALUE,
      FSN_ADT_NOTES_VALUE,
      BAHMNI_VALUE
    );
    if (getAdtNotes.status === 200) {
      setShowAdtNotes(true);
      setAdtNotesObj(getAdtNotes.data.results[0]);
    }
  };
  useEffect(() => {
    fetchAdtNotes();
  }, []);

  const createAdtNotesObj = () => {
    const notesObj = {
      concept: {
        uuid: adtNotesObj.uuid,
        name: adtNotesObj.name.name,
        dataType: adtNotesObj.datatype.name,
      },
      units: adtNotesObj.units,
      label: adtNotesObj.name.display,
      possibleAnswers: adtNotesObj.answers,
      groupMembers: adtNotesObj.setMembers,
      comment: null,
      isObservation: true,
      conceptUIConfig: [],
      value: adtNotes,
      autocompleteValue: adtNotes,
      conceptSetName: adtNotesObj.name.name,
    };
    return notesObj;
  };
  const createPayload = () => {
    const observations = adtNotes && showAdtNotes ? [createAdtNotesObj()] : [];
    const obj = {
      patientUuid: patient.uuid,
      encounterTypeUuid: selectedDropDownItem.encounterType,
      visitTypeUuid: vistEncounterTypes.visitTypes.IPD,
      observations: observations,
      locationUuid: location,
      providers: [provider],
    };
    return obj;
  };

  const handleOnSave = async () => {
    if (selectedDropDownItem) {
      setSaveEnable(false);
      if (selectedDropDownItem.name === "Admit Patient") {
        const response = await updatePatientMovement(createPayload());
        if (response.status === 200) {
          window.location.href = getADTDashboardUrl(
            patient.uuid,
            visit,
            response.data.encounterUuid
          );
        }
      } else if (selectedDropDownItem.name === "Discharge Patient") {
        const response = await dischargePatient(createPayload());
        if (response.status === 200) {
          updatePatientMovementModal(false);
        }
      } else if (selectedDropDownItem.name === "Undo Discharge") {
        const response = await undoDischargePatient(
          visitSummary.dischargeDetails.uuid,
          selectedDropDownItem.name
        );
        if (response.status === 200) {
          window.location.href = getADTDashboardUrl(
            patient.uuid,
            visit,
            visitSummary.dischargeDetails.uuid
          );
        }
      } else if (selectedDropDownItem.name === "Transfer Patient") {
        const response = await updatePatientMovement(createPayload());
        if (response.status === 200) {
          window.location.href = getADTDashboardUrl(
            patient.uuid,
            visit,
            response.data.encounterUuid
          );
        }
      }
    } else {
      setSaveEnable(true);
    }
  };

  const handleSelectOnChange = (e) => {
    var item = {};
    setCurrentItem(e);
    setSaveEnable(false);
    if (e.label === "Admit Patient") {
      item = {
        name: "Admit Patient",
        encounterType: vistEncounterTypes.encounterTypes.ADMISSION,
      };
      setShowAdtNotes(true);
    } else if (e.label === "Undo Discharge") {
      item = { name: "Undo Discharge", encounterType: null };
      setShowAdtNotes(false);
    } else if (e.label === "Discharge Patient") {
      item = {
        name: "Discharge Patient",
        encounterType: vistEncounterTypes.encounterTypes.DISCHARGE,
      };
      setShowAdtNotes(true);
    } else if (e.label === "Transfer Patient") {
      item = {
        name: "Transfer Patient",
        encounterType: vistEncounterTypes.encounterTypes.TRANSFER,
      };
      setShowAdtNotes(false);
    } else {
      setSaveEnable(true);
    }
    setSelectedDropDownItem(item);
  };

  const handleNotesOnChange = (e) => {
    setAdtNotes(e.target.value);
  };

  useEffect(() => {
    if (currentItem != {}) {
      setSaveEnable(false);
    }
    else {
      setSaveEnable(true);
    }
  },[currentItem]);
  
  return (
    <>
      {!isLoading && (
        <Modal
          open
          modalHeading="Patient Movement"
          primaryButtonText="Save"
          secondaryButtonText="Cancel"
          aria-label="Modal content"
          onRequestClose={() => updatePatientMovementModal(false)}
          onSecondarySubmit={() => updatePatientMovementModal(false)}
          onRequestSubmit={() => handleOnSave()}
          primaryButtonDisabled={saveEnable}
        >
          
          {dropDown == {} ? (
            <DropdownSkeleton />
          ) : (
            <Dropdown
              id="patient-movement-dropdown"
              placeholder={"Choose the patient movement"}
              titleText={""}
              onChange={(e) => handleSelectOnChange(e.selectedItem)}
              isRequired={true}
              width={"100%"}
              items={dropDown}
              itemToString={(item) => (item ? item.label : "")}
              selectedItem={currentItem}
            />
          )}
          &nbsp;&nbsp;
          {!showAdtNotes && (<div style={{height:"50px"}}></div>)}
          {showAdtNotes && (
            <TextArea
              data-modal-primary-focus
              id="ipd-adt-notes"
              placeholder={"Enter " + adtNotesObj.name.display}
              style={{
                marginBottom: "1rem",
              }}
              onChange={(e) => handleNotesOnChange(e)}
            />
          )}
        </Modal>
      )}
    </>
  );
};

export default PatientMovementModal;

PatientMovementModal.propTypes = {
  updatePatientMovementModal: PropTypes.func.isRequired
}
