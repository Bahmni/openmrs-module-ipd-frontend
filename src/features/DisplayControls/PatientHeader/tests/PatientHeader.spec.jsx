import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import React from "react";
import { PatientHeader } from "../components/PatientHeader";




const mockFetchPatientProfile = jest.fn();
const mockContactDetailsConfig  = jest.fn();
const mockFetchAddressMapping = jest.fn();

jest.mock("../utils/PatientHeaderUtils", () => {
  const originalModule = jest.requireActual("../utils/PatientHeaderUtils");
  return {
    ...originalModule,
    fetchPatientProfile: () => mockFetchPatientProfile("123"),
    getConfigsForPatientContactDetails: () => mockContactDetailsConfig(),
    fetchAddressMapping: () => mockFetchAddressMapping()

  };
});


describe("PatientHeader", () => {
  beforeEach(() => {


    mockContactDetailsConfig.mockResolvedValue({
      contactDetails: [
        "residencePhone",
        "mobilePhone",
        "workPhone",
        "otherPhone",
        "email"
    ]
    });
    mockFetchPatientProfile.mockResolvedValue({
     patient: {
       person: {
        preferredName: {
          display: "John Doe",
        },
        age: 30,
        birthdate: "1991-01-01",
        gender: "M",
      "preferredAddress": {
          "display": "Region test",
          "uuid": "a59ae106-3096-49eb-92fe-e3d83c5b84a9",
          "preferred": true,
          "address1": "Region test",
          "address2": "zone test",
          "cityVillage": null,
          "stateProvince": null,
          "country": "Ethiopia",
          "postalCode": null,
          "countyDistrict": null,
          "address3": "W",
          "address4": "Kebele",
          "address5": "13-899",
          "address6": null,
          "startDate": null,
          "endDate": null,
          "latitude": null,
          "longitude": null,
          "voided": false,
          "address7": null,
          "address8": null,
          "address9": null,
          "address10": null,
          "address11": null,
          "address12": null,
          "address13": null,
          "address14": null,
          "address15": null,
      },
      "attributes": [
          {
              "display": "residencePhone = 22222222",
              "uuid": "d462d1c3-6c8c-46e6-bb61-552167455452",
              "value": "22222222",
              "attributeType": {
                  "uuid": "88b2f4fb-65cc-44e5-9c05-e82e8487f65c",
                  "display": "residencePhone",
              },
          },
          {
              "display": "mobilePhone = 999999999",
              "uuid": "1b391322-0694-40d1-923b-95b3b63a15b6",
              "value": "999999999",
              "attributeType": {
                  "uuid": "708b8e27-f7fc-479b-b3cd-f25fe8de0d9c",
                  "display": "mobilePhone",
              },
          },
          {
              "display": "workPhone = 34567890",
              "uuid": "27a67227-bea9-4a84-be67-e4511d63f005",
              "value": "34567890",
              "attributeType": {
                  "uuid": "f8066909-8d23-4914-b55e-fccfb62e4440",
                  "display": "workPhone",
              },
          },
          {
              "display": "otherPhone = 99999999",
              "uuid": "95599a79-886c-4ff8-a0fd-3bd6c7680f4e",
              "value": "99999999",
              "attributeType": {
                  "uuid": "d43bacd2-8162-48e3-927f-e159c308b466",
                  "display": "otherPhone",
              },
          },
          {
              "display": "email = samuel email",
              "uuid": "1e3d7fdb-f4aa-4903-913a-592b18477383",
              "value": "samuel email",
              "attributeType": {
                  "uuid": "fe7f70a1-6c4d-4e2f-b435-f79884fd0031",
                  "display": "email",
              },
          },
        ]},
        identifiers: [
          {
            identifier: "12345",
          },
        ] 
     },
     relationships: [
      {
        "uuid": "5a0a7da5-128f-4311-a39d-d1a2a8eda448",
        "display": "Samuel is the Father of Test",
        "relationshipType": {
          "uuid": "c86d9979-b8ac-4d8c-85cf-cc04e7f16001",
          "display": "Father/Patient"
        },
        "personB": {
          "uuid": "698f8a07-9bb1-429a-a450-5b9867cf80b2",
          "display": "Test Father"
        }
      },
      {
        "uuid": "48aa82ca-3711-40ed-b0f1-bc05013ea803",
        "display": "Samuel is the Mother of Mother",
        "relationshipType": {
          "uuid": "c86d9979-b8ac-4d8c-85cf-cc04e7f16002",
          "display": "Mother/Patient"
        },
        "personB": {
          "uuid": "c732b222-6b9c-494b-ab3d-faea648e004c",
          "display": "Mother Test"
        }
      }
    ]
    })
   mockFetchAddressMapping.mockResolvedValue(
    [
      {
          "name": "Country",
          "addressField": "country",
          "required": false
      },
      {
          "name": "Region",
          "addressField": "address1",
          "required": false
      },
      {
          "name": "Zone",
          "addressField": "address2",
          "required": false
      },
      {
          "name": "Woreda",
          "addressField": "address3",
          "required": false
      },
      {
          "name": "Kebele",
          "addressField": "address4",
          "required": false
      },
      {
          "name": "Address",
          "addressField": "address5",
          "required": false
      }
  ]) 
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it("should render without crashing", () => {
    render(<PatientHeader patientId="123" />);
  });

  it("should call fetchPatientInfo on mount", () => {
    render(<PatientHeader patientId="123" />);
    expect(mockFetchPatientProfile).toHaveBeenCalledWith("123");
  });

  it("should display loading skeleton while fetching data", () => {
    render(<PatientHeader patientId="123" />);
    expect(screen.getByTestId("header-loading")).toBeTruthy();
  });

  it("should display patient details after data is fetched", async () => {
    render(<PatientHeader patientId="123" />);
    await waitFor(() => expect(screen.getByText("John Doe")).toBeTruthy());
    expect(screen.getByText(/30 Years/i)).toBeTruthy();
    expect(screen.getByText("01/01/1991")).toBeTruthy();
    expect(screen.getByText(/12345/i)).toBeTruthy();
  });

  it("should display all details of the patient", async () => {
     const {container} = render(<PatientHeader patientId="123"/>);
    await waitFor(() => 
    expect(screen.getByText("John Doe")).toBeTruthy());
    const showDetailsButton = screen.getByText("Show Details");
    fireEvent.click(showDetailsButton);

    await waitFor (() =>
    {
      expect(screen.getByText(/Country : Ethiopia/i)).toBeTruthy();
      expect(container).toMatchSnapshot();
    })
    // expect(screen.getByText(/Residence Phone : 22222222/i)).toBeTruthy();
    // expect(screen.getByText(/Kebele : Kebele/i)).toBeTruthy();
    // expect(screen.getByText(/Father : Test Father/i)).toBeTruthy();
    // expect(screen.getByText(/Email : samuel email/i)).toBeTruthy();
    // expect(screen.getByText(/Mother : Mother Test/i)).toBeTruthy();

  } )
});
