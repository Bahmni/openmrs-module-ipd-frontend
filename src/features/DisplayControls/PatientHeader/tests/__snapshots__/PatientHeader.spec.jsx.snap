// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatientHeader should display all details of the patient 1`] = `
<div>
  <div
    class="bx--tile patient-header"
  >
    <div
      class="bx--grid"
    >
      <div
        class="patient-image-and-details bx--row"
      >
        <div
          class="patient-image"
        />
        <div
          class="bx--col"
        >
          <div
            class="bx--row"
          >
            <div
              class="bx--col"
            >
              <h1
                class="patient-name"
              >
                John Doe
              </h1>
            </div>
          </div>
          <div
            class="bx--row"
          >
            <div
              class="other-info"
            >
              <div
                class="patient-basic-info"
              >
                <h3
                  class="patient-info"
                >
                  Male
                </h3>
                <h3
                  class="patient-info"
                >
                  30
                   
                  Years
                </h3>
                <h3
                  class="patient-info"
                >
                  01/01/1991
                </h3>
                <h3
                  class="patient-info"
                >
                  12345
                </h3>
              </div>
              <a
                class="bx--link show-more bx--link--sm"
                kind="tertiary"
              >
                Hide Details
                 
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="20"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 32 32"
                  width="20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16 10L26 20 24.6 21.4 16 12.8 7.4 21.4 6 20z"
                  />
                </svg>
                 
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="bx--tile patient-details"
    >
      <div
        class="bx--row"
      >
        <div
          class="bx--col"
        >
          <div
            class="bx--tile details-tile"
          >
            <span
              class="details-headers"
            >
              Address
            </span>
            <span
              class="details-value"
            >
              Country
               : 
              Ethiopia
            </span>
            <span
              class="details-value"
            >
              Region
               : 
              Region test
            </span>
            <span
              class="details-value"
            >
              Zone
               : 
              zone test
            </span>
            <span
              class="details-value"
            >
              Woreda
               : 
              W
            </span>
            <span
              class="details-value"
            >
              Kebele
               : 
              Kebele
            </span>
            <span
              class="details-value"
            >
              Address
               : 
              13-899
            </span>
          </div>
        </div>
        <div
          class="bx--col"
        >
          <div
            class="bx--tile details-tile"
          >
            <span
              class="details-headers"
            >
              Contact Details
            </span>
            <span
              class="details-value"
            >
              Residence Phone
               : 
              22222222
            </span>
            <span
              class="details-value"
            >
              Mobile Phone
               : 
              999999999
            </span>
            <span
              class="details-value"
            >
              Work Phone
               : 
              34567890
            </span>
            <span
              class="details-value"
            >
              Other Phone
               : 
              99999999
            </span>
            <span
              class="details-value"
            >
              Email
               : 
              samuel email
            </span>
          </div>
        </div>
        <div
          class="bx--col"
        >
          <div
            class="bx--tile details-tile"
          >
            <span
              class="details-headers"
            >
              Relationships
            </span>
            <span
              class="details-value"
            >
              Father
               : 
              Test Father
            </span>
            <span
              class="details-value"
            >
              Mother
               : 
              Mother Test
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
